keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,424 @@
+   "msg": "Pod status updated",
+   "pod":{
+      "name": "nginx-1",
+      "namespace": "default"
+   },
+   "status": "ready"
+}
+```
+
+
+### Logging configuration
+
+To allow selection of the logging output format we would like to introduce a new logging configuration shared by all kubernetes components.
+
+`LoggingConfig` structure should be implemented as part of `k8s.io/component-base` options and include a common set of flags for logger initialization.
+
+`--logging-format` flag should allow you to pick between logging logging output formats. Setting this flag will select a particular logger implementation.
+
+Proposed flag `--logging-format` values:
+* `text` for old text-based logging format
+* `json` for new JSON format
+
+### Migration / Graduation Criteria
+
+#### Alpha
+
+Preparation and migration of selected logs to new logging methods:
+* Support for the new InfoS, WarningS, ErrorS and FatalS method will be implemented in klog and the new version of klog will be released.
+* Kubernetes will be upgraded to use the new klog version.
+* Helper methods for producing consistent Kubernetes resource identifiers to be used in logs will be implemented.
+* Most important klog calls will be identified and will be manually migrated from using format based to structured methods.
+
+#### Beta
+
+Adding support for JSON output format and guarding against regression:
+* klog version used by Kubernetes and all of its dependencies will be upgraded to v2.
+* Implementation of logr which will produce logs in JSON format will be created.
+* Flag which will inject new logr implementation into klog if --logging-format is set to json will be added to k8s.io/component-base JSON format
+* Static analysis protects against adding new string formatting logs
+
+#### GA
+
+* All feedback for text and json logs output formats is addressed and both output formats become an API.
+
+### Performance
+
+Proposed changes could impact k8s performance in two ways:
+* By introducing degrading in klog implementation
+* By largely increasing log volume
+
+#### Logger implementation performance
+
+We have benchmarked proof of concept implementation of new logging format. Results measured on 30s benchmark for passing 2 arguments to format function.
+
+|logger                 |time [ns/op]|bytes[B/op]|allocations[alloc/op]|
+|-----------------------|------------|-----------|---------------------|
+|Text Infof             |2252        |248        |3                    |
+|Text InfoS             |2455        |280        |3                    |
+|JSON Infof             |1406        |19         |1                    |
+|JSON InfoS             |319         |67         |1                    |
+
+InfoS implementation for text is 9% slower than Infof. This increase should not have a big impact on overall Kubernetes performance as logging takes less than 2% of overall CPU usage.
+JSON implementation using zapr is much more efficient compared to current implementation.
```

Comment:
I think this doesn't matter unless there's an easy way to switch to zapr?

Reply:
+1
We should compare it with the klog-based implementation that we're going to use IIUC.

Reply:
In this KEP we are proposing to have `json` as alternative format to `text`. It will not be default, still I added performance analysis fot it.